File: sdk/typescript/src/pkce.ts
/**
 * PKCE (Proof Key for Code Exchange) implementation.
 * 
 * OAuth 2.1 compliant implementation that always uses the S256 challenge method.
 * The plain method is intentionally not supported as it provides no security benefit.
 * 
 * @see {@link https://datatracker.ietf.org/doc/html/rfc7636 | RFC 7636 - PKCE}
 * @packageDocumentation
 */

import type { CodeVerifier, CodeChallenge } from './types/branded.js';
import { createCodeVerifier, createCodeChallenge } from './types/branded.js';
import { base64UrlEncode } from './utils/base64url.js';
import { sha256, timingSafeEqual } from './utils/crypto.js';

/**
 * PKCE challenge pair containing verifier and challenge.
 * 
 * The verifier is kept secret and sent during token exchange.
 * The challenge is sent during authorization request.
 */
export interface PKCEChallenge {
  /** Secret code verifier (43-128 characters) */
  readonly codeVerifier: CodeVerifier;
  /** SHA-256 hash of verifier, base64url encoded */
  readonly codeChallenge: CodeChallenge;
  /** Challenge method (always 'S256' for security) */
  readonly codeChallengeMethod: 'S256';
}

/**
 * Generate a cryptographically random code verifier.
 * 
 * Per RFC 7636, the verifier must be 43-128 characters from the
 * unreserved character set [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~".
 * 
 * @returns Branded CodeVerifier string
 * @internal
 */
function generateCodeVerifier(): CodeVerifier {
  const array = new Uint8Array(32);
  crypto.getRandomValues(array);
  const encoded = base64UrlEncode(array);
  return createCodeVerifier(encoded);
}

/**
 * Generate code challenge from verifier using SHA-256.
 * 
 * @param verifier - Code verifier to hash
 * @returns Base64url-encoded SHA-256 hash
 * @internal
 */
async function generateCodeChallengeFromVerifier(
  verifier: CodeVerifier
): Promise<CodeChallenge> {
  const digest = await sha256(verifier);
  const encoded = base64UrlEncode(digest);
  return createCodeChallenge(encoded);
}

/**
 * Generate a PKCE challenge pair for OAuth 2.1 authorization.
 * 
 * Always uses the S256 (SHA-256) challenge method as required by OAuth 2.1.
 * The plain method is not supported for security reasons.
 * 
 * @returns PKCE challenge containing verifier, challenge, and method
 * 
 * @example
 * ```typescript
 * const pkce = await generatePKCE();
 * 
 * // Use challenge in authorization request
 * const authUrl = new URL('https://auth.example.com/authorize');
 * authUrl.searchParams.set('code_challenge', pkce.codeChallenge);
 * authUrl.searchParams.set('code_challenge_method', pkce.codeChallengeMethod);
 * 
 * // Store verifier for token exchange
 * sessionStorage.setItem('pkce_verifier', pkce.codeVerifier);
 * ```
 */
export async function generatePKCE(): Promise<PKCEChallenge> {
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = await generateCodeChallengeFromVerifier(codeVerifier);

  return {
    codeVerifier,
    codeChallenge,
    codeChallengeMethod: 'S256',
  };
}

/**
 * Verify that a code verifier matches a code challenge.
 * 
 * Uses timing-safe comparison to prevent timing attacks.
 * This is typically used server-side during token exchange.
 * 
 * @param codeVerifier - The original code verifier
 * @param codeChallenge - The challenge to verify against
 * @returns `true` if verifier matches challenge, `false` otherwise
 * 
 * @example
 * ```typescript
 * const isValid = await verifyPKCE(codeVerifier, codeChallenge);
 * if (!isValid) {
 *   throw new Error('PKCE verification failed');
 * }
 * ```
 */
export async function verifyPKCE(
  codeVerifier: CodeVerifier,
  codeChallenge: CodeChallenge
): Promise<boolean> {
  const computed = await generateCodeChallengeFromVerifier(codeVerifier);
  return timingSafeEqual(computed, codeChallenge);
}

File: services/session-identity/lib/session_identity_core/oauth/pkce.ex
defmodule SessionIdentityCore.OAuth.PKCE do
  @moduledoc """
  PKCE (Proof Key for Code Exchange) implementation per OAuth 2.1.
  
  Supports S256 method ONLY as required by OAuth 2.1/RFC 9700.
  Uses constant-time comparison from AuthPlatform.Security.
  
  ## Security
  
  - Plain method is rejected (OAuth 2.1 requirement)
  - Code verifier must be 43-128 characters
  - Valid charset: [A-Za-z0-9\\-._~]
  - Uses constant-time comparison to prevent timing attacks
  """

  alias AuthPlatform.Security
  alias SessionIdentityCore.Shared.Errors

  @verifier_min_length 43
  @verifier_max_length 128
  @challenge_length 43
  @valid_chars ~r/^[A-Za-z0-9\-._~]+$/
  @valid_base64url ~r/^[A-Za-z0-9_-]+$/

  @doc """
  Verifies that the code_verifier matches the code_challenge using S256 method.
  
  S256: BASE64URL(SHA256(code_verifier)) == code_challenge
  
  Uses constant-time comparison to prevent timing attacks.
  """
  @spec verify(String.t(), String.t(), String.t()) :: :ok | {:error, atom()}
  def verify(code_verifier, code_challenge, "S256") do
    computed = compute_s256_challenge(code_verifier)

    if Security.constant_time_compare(computed, code_challenge) do
      :ok
    else
      Errors.invalid_code_verifier()
    end
  end

  def verify(_code_verifier, _code_challenge, "plain") do
    # Plain method is NOT allowed per OAuth 2.1
    {:error, :plain_method_not_allowed}
  end

  def verify(_code_verifier, _code_challenge, _method) do
    Errors.unsupported_pkce_method()
  end

  @doc """
  Computes the S256 challenge from a code_verifier.
  
  S256 = BASE64URL(SHA256(code_verifier))
  """
  @spec compute_s256_challenge(String.t()) :: String.t()
  def compute_s256_challenge(code_verifier) do
    :crypto.hash(:sha256, code_verifier)
    |> Base.url_encode64(padding: false)
  end

  @doc """
  Validates code_verifier format per RFC 7636.
  
  Must be 43-128 characters, using only [A-Z], [a-z], [0-9], "-", ".", "_", "~"
  """
  @spec validate_code_verifier(String.t()) :: :ok | {:error, atom()}
  def validate_code_verifier(verifier) when is_binary(verifier) do
    len = String.length(verifier)

    cond do
      len < @verifier_min_length -> Errors.code_verifier_too_short()
      len > @verifier_max_length -> Errors.code_verifier_too_long()
      not Regex.match?(@valid_chars, verifier) -> {:error, :invalid_characters}
      true -> :ok
    end
  end

  def validate_code_verifier(_), do: Errors.invalid_code_verifier()

  @doc """
  Validates code_challenge format.
  
  Must be 43 characters (BASE64URL encoded SHA256 without padding)
  """
  @spec validate_code_challenge(String.t()) :: :ok | {:error, atom()}
  def validate_code_challenge(challenge) when is_binary(challenge) do
    if String.length(challenge) == @challenge_length and Regex.match?(@valid_base64url, challenge) do
      :ok
    else
      Errors.invalid_code_challenge()
    end
  end

  def validate_code_challenge(_), do: Errors.invalid_code_challenge()

  @doc """
  Generates a cryptographically secure code_verifier.
  
  Returns a 43-character URL-safe base64 encoded string.
  """
  @spec generate_code_verifier() :: String.t()
  def generate_code_verifier do
    Security.generate_token(32, encoding: :url_safe_base64)
  end

  @doc """
  Generates a code_challenge from a code_verifier using S256.
  """
  @spec generate_code_challenge(String.t()) :: String.t()
  def generate_code_challenge(code_verifier) do
    compute_s256_challenge(code_verifier)
  end
end
